powerquery-docs/data-privacy-firewall.md

```diff
@@ -154,9 +154,10 @@ Here's a high-level summary of the partitioning logic.
       * Note that "removing" a partition effectively includes it in whatever other partitions reference it.
       * Trimming parameter partitions allows parameter references used within data source function calls (for example, `Web.Contents(myUrl)`) to work, instead of throwing "partition can't reference data sources and other steps" errors.
     * Grouping (Static)
-      * Partitions are merged, while maintaining separation between:
-        * Partitions in different queries
-        * Partitions that reference other partitions vs. those that don't
+      * Partitions will only be merged if both are in the same query.
```

Comment:
How about:

* Partitions are merged using the following rules:
  * A partition can only be merged into a partition it references within the same query (in other words, into its "predecessor").
  * A partition that _only_ references its predecessor can always be merged into that predecessor.
  * A partition that references a partition other than its predecessor (such as multiple partitions within the same query, or a partition in another query) can only be merged into a partition that also references a non-predecessor partition.

Reply:
Thanks, @ehrenMSFT.

Thinking about how a partition can have multiple predecessors. Does this help clarify that case?

* Partitions are merged using the following rules:
  * A partition will only be merged into a partition it references within the same query (which we'll call a "predecessor").
  * A partition that only references a single predecessor is always be merged into that predecessor.
  * A partition that references more than one predecessor (so references multiple partitions in the same query) or that also references a partition other than a predecessor (so also references a partition in another query) will be merged into a predecessor if that predecessor also references multiple predecessors and/or a non-predecessor partition.

Reply:
I took a look at the code. The logic is a bit more subtle than what I wrote above. But I think a full explanation of it is probably overkill for this doc.

What about something like this, which focuses more on the end goal?

* Partitions are merged, while ensuring the resulting partitions don't violate the firewall's basic rules. In the resulting merged partitions, the following will be separate:
  * Partitions in different queries
  * Partitions that don't reference other partitions, and thus are likely to access a data source
  * Partitions that reference other partitions, and thus aren't allowed to access a data source

These separations don't apply to what gets merged, just to what the results of merging ought to look like.

The full logic is more like this:

* Partitions are first assigned a category:
  * Source: A partition which doesn't reference any other partitions
  * Neutral: A partition which references only one partition within the same query
  * QueryOrMultiStepReference: A partition that references a partition in another query, or multiple steps within the same query
* Partitions are then merged in bottom-up dependency order. Partitions A and B will be merged if the following conditions apply:
  * A and B are in the same query.
  * A references B.
  * No other partitions reference B.
  * B is not the top-level result returned by the query.
  * One of the following conditions is true:
    * The categories of A and B are the same, or one of them is Neutral.
    * A is QueryOrMultiStepReference, but only references Source partitions within the same query that can all be merged with it according to the rules above.

Reply:
Thank you, @ehrenMSFT! 

I like the idea of an end goals summary.

While thinking through the full logic description and how it is reflected in the goal summary, I wanted to be sure I understand a rule from the full logic:

> B is not the top-level result returned by the query.

Is this to keep code that's being used segmented from code that is not being used. So, below, to keep B (which shouldn't be evaluated due to M's laziness) from being combined into a partition with A?

````
let
  Source = ...,
  A = Table.Something(Source),
  B = Table.SomethingElse(A)
in
  A
````

Reply:
It's to prevent A's partition from being merged into B's, since A's partition is needed to return the final result of the query.

Reply:
But would the only case that would be encountered be when B is unused to output results?